# Stats 21: Homework 1

## Julie Yang

Acknowledgements: several of these problems are copied from or modified from Think Python by Allen Downey.

I've started the homework file for you. You'll need to fill in the rest with your answers.
My encouragement is to use the keyboard shortcuts as much as possible and use the mouse as little as possible while working the Jupyter Notebook.

After you complete the homework with your answers, go to the menu and choose **Kernel > Restart & Run All.** Review the document to **make sure all requested output is visible**. You will not get credit for problems where the requested output is not visible, even if the function you coded is correct.

When you are satisfied with the output, choose File > Download As ... > PDF or  HTML. If you choose to save as HTML, you'll then need to "Print as PDF". Submit the PDF to Gradescope.

Submit this ipynb file, complete with your answers and output to Canvas / Bruin Learn.

**Again, you must make sure all requested output is visible to receive full credit.**

# Task 1

Create an account on GitHub. 

Change your profile picture. Ideally, use photo of yourself that would be appropriate for a resume. If you are not comfortable with the idea of using a photo of yourself, use any other image that is suitable for a workplace environment.

Follow the instructions provided in class to fork the class repository to your GitHub.

Create another repository with at least two text files in it on GitHub (other than the forked class notes repository). Make at least two additional commits to the reposiotry and push them to GitHub.

Provide a link to both repositories here.

You will also need to submit the link to your own repository (not the forked one) to Canvas / Bruin Learn.

## Your Answer:

- Link to your forked repository: https://github.com/julieyang12/2022-fa-stats21
- Link to your own repository: https://github.com/julieyang12/HW-1

# Problem 2

An important part of programming is learning to interpret error messages and understanding what correction needs to be made.

Read and familiarize yourself with the following error messages.

Explain the error. Then duplicate each cell and correct the error. The first problem has been done for you as an example.

In [1]:
# A
print("Hello World"

SyntaxError: unexpected EOF while parsing (1427017300.py, line 2)

Answer: The `print()` function is missing the closing parenthesis. This results in an unexpected EOF error.

In [2]:
# corrected:
print("Hello World")

Hello World


In [3]:
# B
print("Hello")
    print("Goodbye")

IndentationError: unexpected indent (66443029.py, line 3)

In [4]:
# corrected:
print("Hello")
print("Goodbye")

Hello
Goodbye


Answer: print("Goodbye") should not be indented. This results in an unexpected indent error.

In [5]:
# C
x = 10
if x > 8
    print("x is greater than 8")

SyntaxError: invalid syntax (4211757722.py, line 3)

In [6]:
# corrected:
x = 10
if x > 8:
    print("x is greater than 8")

x is greater than 8


Answer: The if statement needs a colon at the end. This results in an invalid syntax error. 

In [7]:
# D
if x = 10:
    print("x is equal to 10")

SyntaxError: invalid syntax (2686327292.py, line 2)

In [8]:
# corrected:
if x == 10:
    print("x is equal to 10")

x is equal to 10


Answer: The logical operator for "equals" is == not =. This results in an invalid syntax error.

In [9]:
# E
x = 5
if x == 5:
print("x is five")

IndentationError: expected an indented block (2346592504.py, line 4)

In [10]:
# corrected:
x = 5
if x == 5:
    print("x is five")

x is five


Answer: Statements made inside if statements must be indented. This results in an expected an indented block error

In [11]:
# F
l = [1, 2, 50, 10]
l = sort(l)

NameError: name 'sort' is not defined

In [12]:
# corrected:
l = [1, 2, 50, 10]
l.sort()
l

[1, 2, 10, 50]

Answer: The function sort() does not exist. This results in a name is not defined error. The correct function that modifies and sorts an object is .sort(). 

# Problem 3

Use Python as a calculator. Enter the appropriate calculation in a cell and be sure the output value is visible.

A. How many seconds are there in 42 minutes 42 seconds?

In [13]:
(42 * 60) + 42

2562

There are 2,562 seconds in 42 minutes and 42 seconds.

B. There are 1.61 kilometers in a mile. How many miles are there in 10 kilometers?

In [14]:
10 / 1.61

6.211180124223602

There are 6.21 miles in 10 kilometers.

C. If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average 1-mile pace (time to complete 1 mile in minutes and seconds)? What is your average speed in
miles per hour?

In [15]:
2562 / (10 / 1.61) # seconds to complete 1 mile
minutes_seconds = [412.482 // 60, 412.482 % 60]
minutes_seconds

[6.0, 52.48200000000003]

The average 1-mile pace is 6 minutes and 52 seconds.

In [16]:
60 / 6.8667 

8.737821661059899

The average speed in miles per hour is 8.74.

# Problem 4
Write functions for the following problems.

A. The volume of a sphere with radius r is $$V = \frac{4}{3}\pi r^3$$

Write a function `sphere_volume(r)` that will accept a radius as an argument and return the volume.

- Use the function to find the volume of a sphere with radius 5.
- Use the function to find the volume of a sphere with radius 15.

In [17]:
import math

def sphere_volume(r):
    volume = ((4/3) * math.pi * r**3)
    return volume


In [18]:
sphere_volume(5)

523.5987755982989

In [19]:
sphere_volume(15)

14137.166941154068

B. Suppose the cover price of a book is \\$24.95, but bookstores get a 40\% discount. Shipping costs \\$3 for the first copy and 75 cents for each additional copy.

Write a function `wholesale_cost(books)` that accepts an argument for the number of books and will return the total cost of the books plus shipping.

- Use the function to find the total wholesale cost for 60 copies.
- Use the function to find the total wholesale cost for 10 copies.

In [20]:
def wholesale_cost(books):
    cost = (24.95 * books * 0.6) + (3 + (.75 * (books - 1)))
    return cost
    

In [21]:
wholesale_cost(60)

945.4499999999999

In [22]:
wholesale_cost(10)

159.45

C. A person runs several miles. The first and last miles are run at an 'easy' pace. Other than the first and last miles, the other miles are at a faster pace.

Write a function `run_time(miles, warm_pace, fast_pace)` to calculate the time the runner will take. The function accepts three input arguments: how many miles the runner travels (minimum value is 2), the warm-up and cool-down pace, the fast pace. The function will print the time in the format minutes:seconds, and will return a tuple of values: (minutes, seconds)

Use the function to find the time to run a total of 5 miles. The warm-up pace is 8:15 per mile. The speed pace is 7:12 per mile.

Call the function using: `run_time(miles = 5, warm_pace = 495, fast_pace = 432)`

In [23]:
def run_time(miles, warm_pace, fast_pace):
    time_seconds = (2 * warm_pace) + ((miles - 2) * fast_pace)
    minutes = time_seconds // 60
    seconds = time_seconds % 60 
    print(minutes, seconds, sep = ":")
    return minutes, seconds

In [24]:
run_time(miles = 5, warm_pace = 495, fast_pace = 432)

38:6


(38, 6)

Another important skill is to be able to read documentation.

Read the documentation for the function str.split() at https://docs.python.org/3/library/stdtypes.html#str.split

Adjust the function so that the call can be made with minutes and seconds:

`run_time(miles = 5, warm_pace = "8:15", fast_pace = "7:12")`

In [25]:
def run_time(miles, warm_pace, fast_pace):
    warm_split = warm_pace.split(':')
    warm_pace = (int(warm_split[0]) * 60) + int(warm_split[1])
    fast_split = fast_pace.split(':')
    fast_pace = (int(fast_split[0]) * 60) + int(fast_split[1])
    time_seconds = (2 * warm_pace) + ((miles - 2) * fast_pace)
    minutes = time_seconds // 60
    seconds = time_seconds % 60 
    print(minutes, seconds, sep = ":")
    return minutes, seconds

In [26]:
run_time(miles = 5, warm_pace = "8:15", fast_pace = "7:12")

38:6


(38, 6)

# Problem 5

Use `import math` to gain access to the math library.

Create a function `polar(real, imaginary)` that will return the polar coordinates of a complex number.

The input arguments are the real and imaginary components of a complex number. The function will return a tuple of values: the value of the radius `r` and the angle `theta`. 

For a refresher, see: https://ptolemy.berkeley.edu/eecs20/sidebars/complex/polar.html

Show the results for the following complex numbers:

- 1 + i
- -2 - 3i
- 4 + 2i

In [27]:
import math

def polar(real, imaginary):
    r = math.sqrt(real**2 + imaginary**2)
    theta = math.atan2(imaginary, real)
    return r, theta
    

In [28]:
polar(1, 1)

(1.4142135623730951, 0.7853981633974483)

In [29]:
polar(-2, -3)

(3.605551275463989, -2.158798930342464)

In [30]:
polar(4, 2)

(4.47213595499958, 0.4636476090008061)

# Problem 6

Define a function called `insert_into(listname, index, iterable)`. It will accept three arguments, a currently existing list, an index, and another list/tuple that will be inserted at the index position.

Python's built-in function, `list.insert()` can only insert one object.

In [31]:
# write your code here
def insert_into(listname, index, iterable):
    start = listname[0:(index)]
    end = listname[index:(len(listname))]
    full_list = start + iterable + end
    return full_list


In [32]:
# do not modify. We will check this result for grading
l = [0,'a','b','c',4,5,6]
i = ['hello', 'there']
insert_into(l, 3, i)

[0, 'a', 'b', 'hello', 'there', 'c', 4, 5, 6]

# Problem 7

Define a function called `first_equals_last(listname)`

It will accept a list as an argument. It will return `True` if the first and last elements are equal and the if the list has a length greater than 1. It will return False for all other cases.

In [33]:
# write your function here
def first_equals_last(listname):
    if listname[0] == listname[len(listname) - 1] and len(listname) > 1:
        return True
    else:
        return False

In [34]:
# do not modify. We will check this result for grading
a = [1,2,3]
first_equals_last(a)

False

In [35]:
# do not modify. We will check this result for grading
b = ['hello','goodbye','hello']
first_equals_last(b)

True

In [36]:
# do not modify. We will check this result for grading
c  = [1,2,3,'1']
first_equals_last(c)

False

In [37]:
# do not modify. We will check this result for grading
d = [[1,2],[3,2],[1,2]]
first_equals_last(d)

True